src/notebook/notebookKernel.ts

```diff
@@ -0,0 +1,197 @@
+
+    }
+
+    public async start() {
+        if (!this._terminal) {
+            const connectedPromise = new Subject()
+            const serverListeningPromise = new Subject()
+
+            const pn = generatePipeName(uuid(), 'vscjl-nbk')
+
+            const server = net.createServer(socket => {
+                this._msgConnection = createMessageConnection(
+                    new StreamMessageReader(socket),
+                    new StreamMessageWriter(socket)
+                )
+
+                this._msgConnection.onNotification(notifyTypeRunCellSucceeded, ({ request_id }) => {
+                    const edit = new vscode.WorkspaceEdit()
+                    // const runEndTime = Date.now()
+
+                    const request = this.executionRequests.get(request_id)
+
+                    edit.replaceNotebookCellMetadata(request.cell.notebook.uri, request.cell.index, request.cell.metadata.with({
+                        runState: vscode.NotebookCellRunState.Success,
+                        lastRunDuration: 0,
+                    }))
+                    edit.replaceNotebookCellOutput(request.cell.notebook.uri, request.cell.index, [])
+                    vscode.workspace.applyEdit(edit)
+
+                    // cell.metadata.statusMessage = formatDuration(Date.now() - start)
+                    // cell.metadata.runState = output === 'success' ? vscode.NotebookCellRunState.Success : vscode.NotebookCellRunState.Error
+                })
+
+                this._msgConnection.onNotification(notifyTypeRunCellFailed, ({ request_id }) => {
+                    const edit = new vscode.WorkspaceEdit()
+                    // const runEndTime = Date.now()
+
+                    const request = this.executionRequests.get(request_id)
+
+                    edit.replaceNotebookCellMetadata(request.cell.notebook.uri, request.cell.index, request.cell.metadata.with({
+                        runState: vscode.NotebookCellRunState.Error,
+                        lastRunDuration: 0,
+                    }))
+                    vscode.workspace.applyEdit(edit)
+
+                    // cell.metadata.statusMessage = formatDuration(Date.now() - start)
+                    // cell.metadata.runState = output === 'success' ? vscode.NotebookCellRunState.Success : vscode.NotebookCellRunState.Error
+                })
+
+                this._msgConnection.onNotification(notifyTypeDisplay, ({ mimetype, current_request_id, data }) => {
+                    // TODO This doesn't work properly, nothing ever gets shown
+                    const executionRequest = this.executionRequests.get(current_request_id)
+
+                    if (executionRequest) {
+                        const cell = executionRequest.cell
+
+                        const edit = new vscode.WorkspaceEdit()
+                        edit.appendNotebookCellOutput(cell.notebook.uri, cell.index, [new vscode.NotebookCellOutput([new vscode.NotebookCellOutputItem(mimetype, data)])])
```

Comment:
@DonJayamanne I'm doing something wrong here, nothing gets ever shown. If I debug this, this code is called, so I must be making some mistake in how I use the cell editing API.

Reply:
Odd, pretty me check. I just realized there were submodules. And that's why I could not get anything to work.

Reply:
FYI - The API for execution will be changing significantly, we're expecting a lot of changes. I'm not going to focus too much on how we update the outputs and the like, as most will change.
Here's some of the on going work, https://github.com/microsoft/vscode/pull/116416
This API will hopefully be ready early next week, I'll continue with areas that doesn't relate to execution to avoid too much re-work or throw away.

Reply:
I think the problem here is these edits are all async, and we're not chaining them.
We had similar (more messy) problems in Jupyter extension.
I'll submit a PR for this.